drivers/at_cmd/at_cmd.c

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <logging/log.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include <net/socket.h>
+#include <init.h>
+#include <bsd_limits.h>
+
+#include <at_cmd.h>
+
+LOG_MODULE_REGISTER(at_cmd, CONFIG_AT_CMD_LOG_LEVEL);
+
+#define RESPONSE_MAX_LEN  CONFIG_AT_CMD_RESPONSE_MAX_LEN
+#define THREAD_PRIORITY   K_PRIO_PREEMPT(CONFIG_AT_CMD_THREAD_PRIO)
+
+#define AT_CMD_OK_STR    "OK"
+#define AT_CMD_ERROR_STR "ERROR"
+#define AT_CMD_CMS_STR   "+CMS:"
+#define AT_CMD_CME_STR   "+CME:"
+
+static K_THREAD_STACK_DEFINE(socket_thread_stack, 2048);
+K_MSGQ_DEFINE(return_code_msq, sizeof(enum at_cmd_rc), 2, 4);
+static K_MUTEX_DEFINE(cmd_pending);
+
+static int              common_socket_fd;
+static int              last_error;
+static struct k_thread  socket_thread;
+static at_cmd_handler_t received_data_handler;
+static char             *response_buf;
+static u32_t            response_buf_len;
+
+struct callback_work_item {
+	struct k_work work;
+	u32_t         data_len;
+	char          data[RESPONSE_MAX_LEN];
+};
+
+static int open_socket(void)
+{
+	common_socket_fd = socket(AF_LTE, 0, NPROTO_AT);
+
+	if (common_socket_fd == -1) {
+		LOG_ERR("AT socket creation failed");
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+static enum at_cmd_rc get_return_code(char *buf)
+{
+	enum at_cmd_rc return_code = AT_CMD_NO_RETURN_CODE;
+	char *tmpstr = NULL;
+
+	do {
+		tmpstr = strstr(buf, AT_CMD_OK_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_OK;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_ERROR_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_ERROR;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CMS_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_CMS;
+			break;
+		}
+
+		tmpstr = strstr(buf, AT_CMD_CME_STR);
+		if (tmpstr) {
+			return_code = AT_CMD_CME;
+			break;
+		}
+	} while (0);
+
+	if ((return_code == AT_CMD_CMS) || (return_code == AT_CMD_CME)) {
+		last_error = atoi(&buf[strlen(AT_CMD_CMS_STR)]);
```

Comment:
Please use `sizeof() - 1` or create a macro to do that. This length is known at compilation time.

Reply:
ok